feat(tests): make tests more independent

vbudhram opened the pull request:
This PR updates the db tests to use more features of mocha. This is cleanup work that will make working with this repo more manageable.

- Tests can run independently of each other
- Testing blocks are much smaller with same coverage
- Expanded some tests and removed duplicate tests
- Removed checking email on change password token https://github.com/mozilla/fxa-auth-db-mysql/pull/293/commits/9cc30b1ca88547a933d255b7b44dbc658489ffda. Turns out that mysql was not returning this but the mem db was. Looking at the auth-server the email was not being used so I think it is fair to remove from mem.

The easiest way I think this could be reviewed is doing a side by side comparison of original db tests. Sorry in advance for including some styling changes :(

Fixes #246

vbudhram commented:
@mozilla/fxa-devs I think this is ready! Review at leisure.

philbooth commented on db-server/test/backend/db_tests.js:
```diff
@@ -54,1938 +54,1313 @@ function createEmail(data) {
+      beforeEach(() => {
+        accountData = createAccount()
+        return db.accountExists(accountData.emailBuffer)
+          .then(() => assert.fail, (err) => assert.equal(err.code, 404, 'Not found'))
```
One thing I'm wondering is whether these `assert.fail` without an explicit error message will come to annoy us a bit in the future. No need to change them now but if we find they're annoying to debug in practice, we may want to come back to them.

philbooth replied:
Actually, should we even be asserting inside the `beforeEach`? The call to `db.createAccount` is going to fail anyway if the account exists, and that error may be more meaningful to us in the stack trace than a failed assertion with no message?

rfk replied:
IIUC this form won't work correctly anyway:

> .then(() => assert.fail)

I don't think that will actually cause `assert.fail` to be executed, you want either `.then(assert.fail)` or `.then(() => assert.fail("some message"))`

vbudhram replied:
Nice catch, fixed!

philbooth commented on db-server/test/backend/db_tests.js:
```diff
@@ -54,1938 +54,1313 @@ function createEmail(data) {
+
+      describe('db.createAccount', () => {
+        it('should create account', () => {
+          accountData = createAccount()
```
This shouldn't be here, right? The same code is already in the parent `beforeEach`.

philbooth commented on db-server/test/backend/db_tests.js:
```diff
@@ -54,1938 +54,1313 @@ function createEmail(data) {
+      })
+
+      describe('db.account', () => {
+        it('should return account', () => {
```
When there is a `describe` with no `before`/`after`/`beforeEach`/`afterEach`, I tend to think the enclosed tests should just be moved up a level into the parent `describe`. The extra layer of indirection isn't adding any value.

philbooth commented on db-server/test/backend/db_tests.js:
```diff
@@ -54,1938 +54,1313 @@ function createEmail(data) {
+              assert.deepEqual(account.authSalt, accountData.authSalt, 'authSalt')
+              assert.equal(account.verifierVersion, accountData.verifierVersion, 'verifierVersion')
+              assert.equal(account.verifierSetAt, accountData.verifierSetAt, 'verifierSetAt')
+              // locale not returned with .emailRecord() (unlike .account() when it is)
```
Rather than a comment saying this, would an assertion that proves it be more useful?

philbooth commented on db-server/test/backend/db_tests.js:
```diff
@@ -54,1938 +54,1313 @@ function createEmail(data) {
         })
-        .then(function() {
-          return db.account(ACCOUNT.uid)
+    })
```
It looks like every child of this `describe` has `accountData = createAccount()` in its `beforeEach`. Is it worth adding a `beforeEach` at this level and refactoring all of those to one place?

vbudhram commented:
@philbooth Thanks for review, only thing to note is from https://github.com/mozilla/fxa-auth-db-mysql/pull/293#discussion_r162898770, the mem db didn't actually do. I updated that to reflect the assertion. Other than that, updated per comments.
